### PR TITLE
[C++ Worker]Generate a template project for users

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?

We should support safe c++ worker API according the [doc](https://docs.google.com/document/d/1pBRUm-w2LKXTte8uXmvQ23l7gOJV2IOwirCTPjtPGWk/edit#heading=h.86fui9aqjkh6), this pr generate a template project for c++ worker users. The c++ worker user don't need to compile ray, just use the template project which contains c++ worker api headers and ray runtime dynamic library.

[More detail](https://docs.google.com/document/d/1brqAkOPVnTDtRvGJM8-dHFyuGjbTHRSBMOdBU43BG9k/edit?usp=sharing)

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

File: cpp/BUILD.bazel [hunk 1]
```diff
@@ -3,6 +3,18 @@
 
 load("//bazel:ray.bzl", "COPTS")
 
+cc_binary(
+    name = "libray_api.so",
+    copts = COPTS,
+    #linkopts = ["-shared"],
```
Comment (SongGuyang) [inline cpp/BUILD.bazel:9]: Remove this?

File: cpp/BUILD.bazel [hunk 2]
```diff
@@ -60,13 +73,29 @@ genrule(
...
+        "libray_api.so",
     ],
     outs = ["ray_cpp_pkg.out"],
     cmd = """
         WORK_DIR="$$(pwd)" &&
         mkdir -p "$$WORK_DIR/python/ray/core/src/ray/cpp/" &&
         cp -f $(location default_worker) "$$WORK_DIR/python/ray/core/src/ray/cpp/default_worker" &&
         cp -f $(locations ray_api) "$$WORK_DIR/python/ray/core/src/ray/cpp/" &&
+        cp -f $(locations libray_api.so) "$$WORK_DIR/python/ray/core/src/ray/cpp/" &&
+        mkdir -p "$$WORK_DIR/cpp/example/thirdparty/include/" &&
```
Comment (SongGuyang) [inline cpp/BUILD.bazel:85]: Maybe we should copy files in `cpp/example/BUILD.bazel`? `ray_cpp_pkg ` don't know the example program.

File: cpp/BUILD.bazel [hunk 3]
```diff
@@ -60,13 +73,29 @@ genrule(
...
+        mkdir -p "$$WORK_DIR/cpp/example/thirdparty/lib/" &&
+        cp -f "$$WORK_DIR/.bazelrc" "$$WORK_DIR/cpp/example/" &&
+        cp -f -r "$$WORK_DIR/external/msgpack/include/msgpack" "$$WORK_DIR/cpp/example/thirdparty/include" &&
+        cp -f "$$WORK_DIR/external/msgpack/include/msgpack.h" "$$WORK_DIR/cpp/example/thirdparty/include" &&
+        cp -f "$$WORK_DIR/external/msgpack/include/msgpack.hpp" "$$WORK_DIR/cpp/example/thirdparty/include" &&
+        cp -f -r "$$WORK_DIR/external/boost/boost/" "$$WORK_DIR/cpp/example/thirdparty/include" &&
+        cp -f -r "$$WORK_DIR/external/com_google_absl/absl/" "$$WORK_DIR/cpp/example/thirdparty/include" &&
+        rm -rf "$$WORK_DIR/cpp/example/thirdparty/include/absl/BUILD.bazel" &&
```
Comment (SongGuyang) [inline cpp/BUILD.bazel:93]: Can we avoid `rm`? Just copy the files want do you want?

File: cpp/example/BUILD.bazel [hunk 4]
```diff
@@ -11,15 +6,10 @@ cc_binary(
...
-    copts = COPTS,
-    data = [
-        "example.so",
-        "//cpp:ray_cpp_pkg",
-    ],
     linkstatic = True,
     deps = [
-        "//cpp:ray_api",
-        "@com_github_gflags_gflags//:gflags",
+        ":example.so",
```
Comment (SongGuyang) [inline cpp/example/BUILD.bazel:11]: Why deps example.so? Should we move this to `data` block?

File: cpp/include/ray/api/function_manager.h [hunk 5]
```diff
@@ -226,6 +226,17 @@ class FunctionManager {
     return instance;
   }
 
+  std::vector<std::string> GetRemoteFunctionNames() {
```
Comment (SongGuyang) [inline cpp/include/ray/api/function_manager.h:229]: Why need this functon in current PR?

**Comment by kfstorm** [comment]
Is there a doc to describe how users can use this template project?

File: cpp/src/ray/util/function_helper.cc [hunk 6]
```diff
@@ -45,6 +45,19 @@ std::shared_ptr<boost::dll::shared_library> FunctionHelper::LoadDll(
...
+  auto get_names_func =
+      boost::dll::import_alias<std::vector<std::string>()>(lib, "GetRemoteFunctionNames");
+  std::string names_str;
+  auto names = get_names_func();
+  for (const auto &name : names) {
+    names_str.append(name).append(", ");
+  }
+  names_str.pop_back();
+  names_str.pop_back();
+  RAY_LOG(INFO) << "All remote functions: " << names_str;
```
Comment (SongGuyang) [inline cpp/src/ray/util/function_helper.cc:58]: Can we also output the lib name?

**Review by SongGuyang — APPROVED** [review]

**Comment by qicosmos** [comment]
> Is there a doc to describe how users can use this template project?

[More detail](https://docs.google.com/document/d/1brqAkOPVnTDtRvGJM8-dHFyuGjbTHRSBMOdBU43BG9k/edit?usp=sharing)

File: cpp/BUILD.bazel [hunk 7]
```diff
@@ -60,13 +72,24 @@ genrule(
...
         WORK_DIR="$$(pwd)" &&
         mkdir -p "$$WORK_DIR/python/ray/core/src/ray/cpp/" &&
         cp -f $(location default_worker) "$$WORK_DIR/python/ray/core/src/ray/cpp/default_worker" &&
         cp -f $(locations ray_api) "$$WORK_DIR/python/ray/core/src/ray/cpp/" &&
+        cp -f $(locations libray_api.so) "$$WORK_DIR/python/ray/core/src/ray/cpp/" &&
+        TEMP_DIR="$$WORK_DIR/cpp/example" &&
+        rm -rf $$TEMP_DIR/thirdparty $$TEMP_DIR/.bazelrc &&
+        mkdir -p "$$TEMP_DIR/thirdparty/include/" &&
+        mkdir -p "$$TEMP_DIR/thirdparty/lib/" &&
+        cp -f "$$WORK_DIR/.bazelrc" "$$TEMP_DIR" &&
```
Comment (kfstorm) [inline cpp/BUILD.bazel:88]: Is `.bazelrc` required?

File: cpp/example/BUILD.bazel [hunk 8]
```diff
@@ -28,11 +20,25 @@ cc_binary(
...
+cc_library(
+    name = "ray_api",
+    srcs = [
+        "thirdparty/lib/libray_api.so",
     ],
+    hdrs = glob([
+        "thirdparty/include/**/*.h",
+        "thirdparty/include/**/*.hpp",
+        "thirdparty/include/**/**/*.h",
+        "thirdparty/include/**/**/*.hpp",
```
Comment (kfstorm) [inline cpp/example/BUILD.bazel:39]: I think `**` is enough here. Why do we need `**/**`?

**Review by kfstorm — APPROVED** [review]

**Pull request merged**